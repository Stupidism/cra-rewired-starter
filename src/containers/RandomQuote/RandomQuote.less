.RandomQuote {
  font-weight: lighter;
  text-align: left;
  font-style: italic;
  display: flex;
  flex-direction: column;

  & > div:nth-child(2) {
    text-align: right;
  }
}
