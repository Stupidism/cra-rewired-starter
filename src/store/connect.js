import { createConnect } from 'react-redux/lib/connect/connect';
import defaultMapStateToPropsFactories from 'react-redux/lib/connect/mapDispatchToProps';
import { wrapMapToPropsFunc } from 'react-redux/lib/connect/wrapMapToProps';

import _ from 'lodash';

export const createMapStateToPropsFromStateMapper = stateMapper => state =>
  _.mapValues(stateMapper, selector => {
    if (typeof selector === 'function') {
      return selector(state);
    }

    throw new Error(`Invalid selector type ${typeof selector} for stateMapper`);
  });

export const whenMapStateToPropsIsObject = stateMapper => {
  if (typeof stateMapper !== 'object') return undefined;

  const mapStateToProps = createMapStateToPropsFromStateMapper(stateMapper);

  return wrapMapToPropsFunc(mapStateToProps, 'mapStateToProps');
};

const mapStateToPropsFactories = [
  ...defaultMapStateToPropsFactories,
  whenMapStateToPropsIsObject,
];

export default createConnect({ mapStateToPropsFactories });
