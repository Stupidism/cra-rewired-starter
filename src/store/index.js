import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import rootReducer from './reducers';

export const configureStore = (initialState = {}, reducer = state => state) => {
  const enhancers = [];
  const middleware = [thunk];

  const composeEnhancers =
    process.env.NODE_ENV === 'development' ? composeWithDevTools : compose;

  const composedEnhancers = composeEnhancers(
    applyMiddleware(...middleware),
    ...enhancers,
  );

  return createStore(reducer, initialState, composedEnhancers);
};

export default configureStore({}, rootReducer);
