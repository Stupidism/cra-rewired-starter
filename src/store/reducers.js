import { combineReducers } from 'redux';
import runtime from './modules/runtime';

export default combineReducers({
  runtime,
});
