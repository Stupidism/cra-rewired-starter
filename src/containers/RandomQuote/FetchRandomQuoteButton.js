import React from 'react';
import { Button } from 'antd';

import connect from 'store/connect';
import {
  fetchRandomQuote,
  isRandomQuoteLoading,
} from 'store/modules/randomQuote';

import './RandomQuote.less';

export const FetchRandomQuoteButton = ({ loading, onLoad, children }) => (
  <Button type="primary" onClick={onLoad} loading={loading}>
    {children}
  </Button>
);

export default connect(
  {
    loading: isRandomQuoteLoading,
  },
  {
    onLoad: fetchRandomQuote,
  },
)(FetchRandomQuoteButton);
