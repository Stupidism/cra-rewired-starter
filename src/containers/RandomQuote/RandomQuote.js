import React from 'react';
import { connect } from 'react-redux';
import {
  compose,
  lifecycle,
  branch,
  renderComponent,
  mapProps,
} from 'recompose';

import { fetchRandomQuote } from 'store/modules/randomQuote';

import FetchRandomQuoteButton from './FetchRandomQuoteButton';

import './RandomQuote.less';

const RandomQuoteLoading = ({ loadingComponent }) => loadingComponent;

const RandomQuoteFailed = () => (
  <div>
    Fail to load random quote.{' '}
    <FetchRandomQuoteButton>Reload</FetchRandomQuoteButton>
  </div>
);

const RandomQuote = ({ quote, author }) => (
  <div className="RandomQuote">
    <div>{quote}</div>
    <div>-- {author}</div>
  </div>
);

export function componentDidMount() {
  if (!this.props.value) {
    this.props.fetchRandomQuote();
  }
}

export default compose(
  connect(state => state.randomQuote, { fetchRandomQuote }),
  branch(({ failed }) => failed, renderComponent(RandomQuoteFailed)),
  branch(({ loading }) => loading, renderComponent(RandomQuoteLoading)),
  lifecycle({ componentDidMount }),
  mapProps(({ value }) => ({ ...value })),
)(RandomQuote);
