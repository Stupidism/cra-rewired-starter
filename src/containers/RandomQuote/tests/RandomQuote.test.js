import React from 'react';
import { mount } from 'enzyme';

import { configureStore } from 'store';

import RandomQuote, { componentDidMount } from '../RandomQuote';

describe('<RandomQuote />', () => {
  describe('should be consistent with snapshot', () => {
    let randomQuoteMock;

    afterEach(() => {
      const storeMock = configureStore({ randomQuote: randomQuoteMock });
      const snapshot = mount(<RandomQuote loadingComponent="Loading..." />, {
        context: {
          store: storeMock,
        },
      });
      expect(snapshot).toMatchSnapshot();
    });

    it('when data fetched', () => {
      randomQuoteMock = {
        value: {
          quote: 'All quotes are stupid except mine',
          author: 'Stupidism',
        },
      };
    });

    it('when still fetching data', () => {
      randomQuoteMock = {
        loading: true,
      };
    });

    it('when failed to fetch data', () => {
      randomQuoteMock = {
        failed: true,
      };
    });
  });

  describe('componentDidMount', () => {
    it('should call fetchRandomQuote', () => {
      const props = {
        value: null,
        fetchRandomQuote: jest.fn(),
      };

      componentDidMount.call({ props });
      expect(props.fetchRandomQuote).toHaveBeenCalledTimes(1);
    });

    it('should not call fetchRandomQuote', () => {
      const props = {
        value: {},
        fetchRandomQuote: jest.fn(),
      };

      componentDidMount.call({ props });
      expect(props.fetchRandomQuote).not.toHaveBeenCalled();
    });
  });
});
