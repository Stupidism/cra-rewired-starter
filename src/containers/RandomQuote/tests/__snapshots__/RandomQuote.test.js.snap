// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RandomQuote /> should be consistent with snapshot when data fetched 1`] = `
<div
  className="RandomQuote"
>
  <div>
    All quotes are stupid except mine
  </div>
  <div>
    -- 
    Stupidism
  </div>
</div>
`;

exports[`<RandomQuote /> should be consistent with snapshot when failed to fetch data 1`] = `
<div>
  Fail to load random quote.
   
  <button
    className="ant-btn ant-btn-primary"
    onClick={[Function]}
    type="button"
  >
    <span
      key=".0"
    >
      Reload
    </span>
  </button>
</div>
`;

exports[`<RandomQuote /> should be consistent with snapshot when still fetching data 1`] = `"Loading..."`;
